modules/caddyhttp/reverseproxy/fastcgi/pool.go

```diff
@@ -0,0 +1,12 @@
+package fastcgi
+
+import (
+	"bytes"
+	"sync"
+)
+
+var bufPool = sync.Pool{
```

Comment:
Maybe it's not a common in this codebase, but what about some type with Acquire and Release methods that is type safe and does the Reset on Release - less bolerplate on use site? Like

```
var bufPool = bufferPool{p: sync.Pool{ New: func() any { return new(bytes.Buffer) } } }
type bufferPool struct {
  p sync.Pool
}
func (p bufferPool) Acquire() *bytes.Buffer { return p.p.Get().(*bytes.Buffer) }
func (p bufferPool) Release(b *bytes.Buffer) { b.Reset(); p.p.Put(b) }
```

Reply:
@tgulacsi The buffer should be reset when _retrieving_ it from the pool, since a buffer that is put back into the pool may not be used again.

Personally, I prefer the vanilla pool style without a wrapper... but that's just me. I don't know if I have a good technical reason though :smile:

Reply:
Are you sure? 
AFAIK a pool is for reuse (minus the GC), so items put may be reused later.

https://pkg.go.dev/sync#Pool says "Pool's purpose is to cache allocated but unused items for later reuse".

So it _is_ for reuse, and newly allocated buffers does not need to Reset, only Put'd ones.


Anyway, by calling Reset before Put, the Get won't need to call Reset, and newly allocated Buffers won't be Reset'd.

Reply:
That same link contains an example in which reset is called after getting buffer from the pool.

Anyway, why bother with the wrapper since there is only one call site?

Reply:
@tgulacsi Yes, I'm sure. From the same godoc:

> If the Pool holds the only reference when this happens, the item might be deallocated.

Reply:
"Any item stored in the Pool may be removed automatically at any time without notification. If the Pool holds the only reference when this happens, the item might be deallocated."

The second part refers to the first part. So this is in effect only when the Pool is evicted.


But if you're not used to use such a type decoration for easier and more error-proof use of a pool, I'll not argue anymore.